File: server/src/metadata/koharu.rs
use crate::{config::CONFIG, db, utils};
use serde::Deserialize;
use slug::slugify;

#[derive(Deserialize)]
pub struct Metadata {
  pub title: String,
  pub description: Option<String>,
  pub source: Option<String>,
  pub artist: Option<Vec<String>>,
  pub circle: Option<Vec<String>>,
  pub parody: Option<Vec<String>>,
  pub magazine: Option<Vec<String>>,
  pub event: Option<Vec<String>>,
  pub general: Option<Vec<String>>,
  pub female: Option<Vec<String>>,
  pub male: Option<Vec<String>>,
  pub mixed: Option<Vec<String>>,
  pub other: Option<Vec<String>>,
  pub language: Option<Vec<String>>,
  pub url: Option<String>,
}

pub fn add_metadata(info: Metadata, archive: &mut db::UpsertArchiveData) -> anyhow::Result<()> {
  archive.title = if CONFIG.metadata.parse_filename_title {
    utils::parse_filename(&info.title).0
  } else {
    Some(info.title)
  };
  archive.slug = archive.title.as_ref().map(slugify);
  archive.description = info.description;
  archive.thumbnail = Some(1);

  archive.artists = info.artist.map(|tags| {
    tags
      .iter()
      .map(|tag| utils::capitalize_words(tag))
      .collect()
  });
  archive.circles = info.circle.map(|tags| {
    tags
      .iter()
      .map(|tag| utils::capitalize_words(tag))
      .collect()
  });
  archive.magazines = info.magazine.map(|tags| {
    tags
      .iter()
      .map(|tag| utils::capitalize_words(tag))
      .collect()
  });
  archive.events = info.event.map(|tags| {
    tags
      .iter()
      .map(|tag| utils::capitalize_words(tag))
      .collect()
  });
  archive.parodies = info.parody.map(|tags| {
    tags
      .iter()
      .map(|tag| utils::capitalize_words(tag))
      .collect()
  });

  archive.language = info.language.and_then(|languages| {
    languages
      .first()
      .map(|language| utils::capitalize_words(language))
  });

  let mut tags = vec![];

  if let Some(general) = info.general {
    general.into_iter().for_each(|tag| {
      tags.push((utils::capitalize_words(&tag), "".to_string()));
    });
  }

  if let Some(female) = info.female {
    female.into_iter().for_each(|tag| {
      tags.push((utils::capitalize_words(&tag), "female".to_string()));
    });
  }

  if let Some(male) = info.male {
    male.into_iter().for_each(|tag| {
      tags.push((utils::capitalize_words(&tag), "male".to_string()));
    });
  }

  if let Some(mixed) = info.mixed {
    mixed.into_iter().for_each(|tag| {
      tags.push((utils::capitalize_words(&tag), "mixed".to_string()));
    });
  }

  if let Some(other) = info.other {
    other.into_iter().for_each(|tag| {
      tags.push((utils::capitalize_words(&tag), "other".to_string()));
    });
  }

  if !tags.is_empty() {
    archive.tags = Some(tags);
  }

  let mut sources = vec![];

  if let Some(url) = info.url {
    sources.push(db::ArchiveSource {
      name: utils::parse_source_name(&url),
      url: Some(url),
    });
  }

  if let Some(url) = info.source {
    sources.push(db::ArchiveSource {
      name: utils::parse_source_name(&url),
      url: Some(format!(
        "https://koharu.to{}",
        url.split(':').last().unwrap()
      )),
    });
  }

  if !sources.is_empty() {
    archive.sources = Some(sources);
  }

  Ok(())
}

File: cli/metadata/koharu.ts
import capitalize from 'capitalize';
import YAML from 'yaml';
import { z } from 'zod';

import config from '../../shared/config';
import { ArchiveMetadata } from '../../shared/metadata';
import { stringOrNumberArray } from './schemas';
import { parseFilename } from './utils';

const metadataSchema = z.object({
	title: z.string(),
	description: z.string().optional(),
	source: z.string().optional(),
	artist: stringOrNumberArray.optional(),
	circle: stringOrNumberArray.optional(),
	parody: stringOrNumberArray.optional(),
	magazine: stringOrNumberArray.optional(),
	event: stringOrNumberArray.optional(),
	general: stringOrNumberArray.optional(),
	female: stringOrNumberArray.optional(),
	male: stringOrNumberArray.optional(),
	character: stringOrNumberArray.optional(),
	mixed: stringOrNumberArray.optional(),
	other: stringOrNumberArray.optional(),
	language: stringOrNumberArray.optional(),
	url: z.string().optional(),
});

export default async (content: string, archive: ArchiveMetadata) => {
	const parsed = YAML.parse(content);
	const { data, error } = metadataSchema.safeParse(parsed);

	if (!data) {
		throw new Error(`Failed to parse Koharu metadata: ${error}`);
	}

	archive = structuredClone(archive);

	if (config.metadata?.parseFilenameAsTitle) {
		archive.title = parseFilename(data.title)[0] ?? data.title;
	} else {
		archive.title = data.title;
	}

	archive.description = data.description;
	archive.language = data.language?.map((language) => capitalize.words(language))[0];

	archive.tags = [];

	if (data.artist) {
		archive.tags.push(...data.artist.map((tag) => ({ namespace: 'artist', name: tag })));
	}

	if (data.circle) {
		archive.tags.push(...data.circle.map((tag) => ({ namespace: 'circle', name: tag })));
	}

	if (data.magazine) {
		archive.tags.push(...data.magazine.map((tag) => ({ namespace: 'magazine', name: tag })));
	}

	if (data.event) {
		archive.tags.push(...data.event.map((tag) => ({ namespace: 'event', name: tag })));
	}

	if (data.parody) {
		archive.tags.push(...data.parody.map((tag) => ({ namespace: 'parody', name: tag })));
	}

	if (data.general) {
		archive.tags.push(...data.general.map((tag) => ({ namespace: 'general', name: tag })));
	}

	if (data.male) {
		archive.tags.push(...data.male.map((tag) => ({ namespace: 'male', name: tag })));
	}

	if (data.female) {
		archive.tags.push(...data.female.map((tag) => ({ namespace: 'female', name: tag })));
	}

	if (data.character) {
		archive.tags.push(...data.character.map((tag) => ({ namespace: 'character', name: tag })));
	}

	if (data.mixed) {
		archive.tags.push(...data.mixed.map((tag) => ({ namespace: 'mixed', name: tag })));
	}

	if (data.other) {
		archive.tags.push(...data.other.map((tag) => ({ namespace: 'other', name: tag })));
	}

	archive.sources = [];

	if (data.url) {
		archive.sources.push({ url: data.url });
	}

	if (data.source) {
		archive.sources.push({ url: `https://koharu.to${data.source.split(':').pop()}` });
	}

	return archive;
};
